url2b521cbda7_eobermuhlner_jhuge_tgz-pJ8-ch_obermuhlner_jhuge_example_MeasureHeapJ8/src/ch/obermuhlner/jhuge/collection/ImmutableHugeHashMap.java

```diff
@@ -0,0 +1,229 @@
+ * <ul>
+ * <li> {@link #put(Object, Object)}</li>
+ * <li> {@link #putAll(Map)}</li>
+ * <li> {@link #remove(Object)}</li>
+ * <li> {@link #clear()}</li>
+ * <li> {@link java.util.Map.Entry#setValue(Object)} on the entries in the {@link #entrySet() entrySet}</li>
+ * <li> {@link Iterator#remove()} when iterating over the {@link #entrySet() entrySet}</li>
+ * </ul>
+ * 
+ * <p>In order to create an {@link ImmutableHugeHashMap} you must add the elements in the {@link Builder}.</p>
+ * 
+ * <p>Other than the mutating operations all semantics, memory consumption and performance are identical to {@link HugeHashMap}.</p>
+ * 
+ * @param <K> the type of keys
+ * @param <V> the type of values
+ * @see HugeHashMap
+ */
+public class ImmutableHugeHashMap<K, V> extends AbstractHugeHashMap<K, V> {
+
+	private int hashCode;
+
+	/**
+	 * Initializes the hashCode.
+	 * 
+	 * Until this method is called the hashCode() and equals() methods are not valid and may not be called.
+	 * Once the Builder is finished with adding all entries he calls initializeHashCode() before this instance is given to the client code.
+	 */
+	private ImmutableHugeHashMap(MemoryManager memoryManager, Converter<K> keyConverter, Converter<V> valueConverter, boolean faster, int capacity) {
+		super(memoryManager, keyConverter, valueConverter, faster, capacity);
+	}
+
+	private void initializeHashCode() {
+		hashCode = super.hashCode();
+	}
+	
+	@Override
+	public int hashCode() {
+		return hashCode;
+	}
+
+	@Override
+	public boolean equals(Object object) {
+		if (object == this) {
+			return true;
+		}
+		if (object instanceof ImmutableHugeArrayList) {
+			// if it is same class as this then we know that hashCode() is cheap (already initialized) -> fast test for not-equal
+			ImmutableHugeHashMap<?, ?> other = (ImmutableHugeHashMap<?, ?>) object;
+			if (hashCode() != other.hashCode()) {
+				return false;
+			}
+		}
+		
+		return super.equals(object);
+	}
+
+	@Override
+	public void clear() {
+		throw new UnsupportedOperationException();
+	}
+	
+	@Override
+	public Set<Entry<K, V>> entrySet() {
+		return new EntrySet();
+	}
+
+	private class EntrySet extends AbstractEntrySet {
+		@Override
+		public Iterator<Entry<K, V>> iterator() {
+			return new EntrySetIterator();
+		}
+	}
+	
+	private class EntrySetIterator extends AbstractEntrySetIterator {
+		@Override
+		public void remove() {
+			throw new UnsupportedOperationException();
+		}
+		
+		@Override
+		protected AbstractEntry createEntry(K key, V value) {
+			return new HugeMapEntry(key, value);
+		}
+	}
+	
+	class HugeMapEntry extends AbstractEntry {
+		public HugeMapEntry(K key, V value) {
+			super(key, value);
+		}
+
+		@Override
+		public V setValue(V value) {
+			return setValueInternal(value);
+		}
+	}
+	
+	/**
+	 * Builds an {@link ImmutableHugeHashMap}.
+	 *
+	 * @param <K> the type of keys
+	 * @param <V> the type of values
+	 */
+	public static class Builder<K, V> extends AbstractHugeMapBuilder<K, V> {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This class contains 11 clusters of instance methods that do not have any accessed class members in common. For example, a cluster might have 2 methods that access only the class fields x and y, and another cluster might have 2 other methods that access only the class fields a and b. By comparison, 98% of the classes in the CodeGuru reference dataset contain fewer clusters. A high number of these clusters indicates low class cohesion. Classes with low class cohesion contain unrelated operations which make them difficult to understand and less likely to be used. We recommend that you simplify this class or break it into multiple cohesive classes.